src/util/threadpool.rs

```diff
@@ -82,21 +82,118 @@ pub trait ScheduleQueue<T: Debug> {
 #[derive(Debug)]
 pub struct ReachConcurrencyLimit<T: Debug>(pub T);
 
+#[derive(Default)]
 struct GroupStatisticsItem {
     running_count: usize,
     high_priority_queue_count: usize,
+    low_priority_queue_count: usize,
 }
 
 impl GroupStatisticsItem {
-    fn new() -> GroupStatisticsItem {
-        GroupStatisticsItem {
-            running_count: 0,
-            high_priority_queue_count: 0,
+    fn sum(&self) -> usize {
+        self.running_count + self.high_priority_queue_count + self.low_priority_queue_count
+    }
+}
+
+// `SmallGroupFirstQueue` tries to run groups with number of
+//  tasks smaller than small_group_tasks_limit first when all threads are busy.
+pub struct SmallGroupFirstQueue<T> {
+    high_priority_queue: VecDeque<Task<T>>,
+    low_priority_queue: VecDeque<Task<T>>,
+    group_concurrency_on_busy: usize,
+    small_group_tasks_limit: usize,
+    statistics: HashMap<T, GroupStatisticsItem>,
+}
+
+impl<T: Hash + Eq + Ord + Send + Clone + Debug> SmallGroupFirstQueue<T> {
+    pub fn new(group_concurrency_on_busy: usize,
+               small_group_tasks_limit: usize)
+               -> SmallGroupFirstQueue<T> {
+        SmallGroupFirstQueue {
+            high_priority_queue: VecDeque::new(),
```

Comment:
no, here `small_group_tasks_limit` is used to determine whether a group is a small group. @andelf